src/lock.rs

```diff
@@ -1,50 +1,77 @@
-use crate::TransactionId;
+use {
+    crate::{AccessKind, TransactionId},
+    smallvec::{smallvec, SmallVec},
+};
 
 /// A read-lock can be held by multiple transactions, and
 /// subsequent write-locks should be blocked by all of them.
 /// Write-locks are exclusive.
-pub(crate) enum Lock<Id: TransactionId> {
-    Read(Vec<Id>, Option<Id>), // (Current Reads, Most Recent Write)
-    Write(Id),
+pub(crate) struct Lock<Id: TransactionId> {
+    kind: AccessKind,
+    // Current Reads or Write
+    current_locks: SmallVec<[Id; 4]>,
```

Comment:
what do you mean by too small to take advantage of?

In terms of perf, one of the big killers (which I should add a bench for) is when we have read-write swaps, because then every insert we are doing an allocation.
`write -> read -> write -> read -> ...`

This basically allows us to keep our allocation there, and will work for up to 4 read-locks before we have an actual allocation.

Reply:
> up to 4 read-locks before we have an actual allocation

This is what I mean - is 4 enough to avoid heap allocation in most cases? I don't have stats, would guess there are more cases take 4+ read locks?